lib/esyBuildEjectCommand.js

```diff
@@ -0,0 +1,234 @@
+    },
+    {
+      type: 'rule',
+      name: '*** Rebuild root package ***',
+      target: 'rebuild',
+      dependencies: [`${packageDb.rootPackageName}.rebuild`],
+    },
+    {
+      type: 'rule',
+      name: '*** Root package shell ***',
+      target: 'shell',
+      dependencies: [`${packageDb.rootPackageName}.shell`],
+    },
+    {
+      type: 'rule',
+      name: '*** Remove sandbox installations / build artifacts ***',
+      target: 'clean',
+      command: 'rm -rf $(ESY__SANDBOX)/_build $(ESY__SANDBOX)/_install',
+    },
+  ];
+
+  let prelude: Array<MakeDefine | MakeRawItem> = [
+    {
+      type: 'raw',
+      value: 'SHELL = /bin/bash',
+    },
+    {
+      type: 'raw',
+      value: 'ESY__SANDBOX ?= $(CURDIR)',
+    },
+    {
+      type: 'define',
+      name: 'ESY__PREPARE_CURRENT_INSTALL_TREE',
+      value: outdent`
+        mkdir -p \\
+          $cur__install \\
+          $cur__lib \\
+          $cur__bin \\
+          $cur__sbin \\
+          $cur__man \\
+          $cur__doc \\
+          $cur__share \\
+          $cur__etc;
+      `,
+    },
+  ];
+
+  traversePackageDb(
+    packageDb,
+    ({normalizedName, packageJsonFilePath, packageJson}) => {
+
+      /**
+       * Produce a package-scoped Makefile rule which executes its command in
+       * the package's environment and working directory.
+       */
+      function makePackageRule(rule: {
+        target: string;
+        dependencies?: Array<string>;
+        command?: ?string;
+      }) {
+        let {
+          target,
+          command,
+          dependencies
+        } = rule;
+        return {
+          type: 'rule',
+          name: `*** ${packageJson.name}: ${target} ***`,
+          target: `${packageJson.name}.${target}`,
+          dependencies,
+          exportEnv: ['SHELL', 'ESY__SANDBOX'],
+          command: command != null
+            ? outdent`
+              $(${normalizedName}__ENV)\\
+              cd $cur__root; \\
+              ${command}
+            `
+            : null
+        };
+      }
+
+      let isRootPackage = packageJson.name === packageDb.rootPackageName;
+      let allDependencies = collectTransitiveDependencies(packageDb, packageJson.name);
+
+      let buildEnvironment = PackageEnvironment.calculateEnvironment(
+        packageDb,
+        packageJson.name
+      );
+
+      // Produce macro with rendered findlib.conf content.
+      let findlibPath = allDependencies
+        .map(dep => installDir(dep, 'lib'))
+        .join(':');
+
+      prelude.push({
+        type: 'define',
+        name: `${normalizedName}__FINDLIB_CONF`,
+        value: outdent`
+          path = "${findlibPath}"
+          destdir = "${installDir(packageJson.name, 'lib')}"
+        `
+      });
+
+      // Produce macro with rendered package's environment.
+      prelude.push({
+        type: 'define',
+        name: `${normalizedName}__ENV`,
+        value: Makefile.renderEnv(buildEnvironment),
+      });
```

Comment:
Right now we dump entire env for a package in a single `define`. Instead we could reduce the size of the `Makefile` by dumping each group into separate `define` and then combine them together for each package. I think this is important to keep the generated `Makefile` readable.